api/types/resource.go

```diff
@@ -145,6 +148,16 @@ func (h *ResourceHeader) SetSubKind(s string) {
 	h.SubKind = s
 }
 
+func (h *ResourceHeader) CheckAndSetDefaults() error {
+	if h.Kind == "" {
+		return trace.BadParameter("resource has an empty Kind field")
+	}
+	if h.Version == "" {
+		return trace.BadParameter("resource has an empty Version field")
+	}
```

Comment:
This is stricter than what tends to be checked with the other resources.

We should perhaps rewrite our proto types to embed the `ResourceHeader` type instead of listing the `Kind`/`SubKind`/`Version` explicitly again and again. Then we could run `CheckAndSetDefaults` on the whole `ResourceHeader` (benefiting from these stricter checks as well) instead of just on the `Metadata` field as it's currently the case.

Reply:
I really want to embed all these common fields and consolidate our API types too!

But I don't think that's an easy change. We need to maintain compatibility of both JSON and proto serialization.
For JSON, embedding tricks would work and keep the resources backwards-compatible.
For protobuf however, serialization is based on the numeric field IDs and embedding won't work.

We can consider a larger migration in the future to solve this, because I agree it causes a lot of code duplication.